src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -158,12 +159,21 @@ public VariableName(VariableName variableName, string childName)
                 }
                 ChildName = childName;
             }
-            public string Name { get; set; }
+            public ROM Name { get; set; }
             public int? Index { get; set; }
-            public string ChildName { get; set; }
-            public override string ToString() => (!Index.HasValue ? Name : Name + "[" + Index.Value + "]") + (ChildName != null ? '.' + ChildName : null);
+            public ROM ChildName { get; set; }
+            public override string ToString() => (!Index.HasValue ? (string)Name : (string)Name + "[" + Index.Value + "]") + (ChildName.Length > 0 ? '.' + (string)ChildName : null);
```

Comment:
@Shane32 Every time you cast ROM to a string, that string is allocated on the heap. That is why I added cached string properties to the AST classes from parser project. And that's why I made minimal edits in my first commit. Your changes are much more extensive, but the main thing is that they do not optimize, but, on the contrary, degrade performance. I myself first followed this path and received dozens of modified files, although at the very beginning I began to suspect something was wrong. I rolled back my changes and started over. I suggest to rollback the last commit. I will run bencmarks on my commit. Perhaps things are not so bad and can be left as it is. Otherwise, I will need to think about how to improve the situation.

Reply:
`ROM.ToString()`:
![изображение](https://user-images.githubusercontent.com/21261007/105179155-02509b80-5b3a-11eb-859c-8b5232d84a0e.png)

Reply:
So `ROM.ToString()) does not allocate **if and only if** it points to `string` and `_index = 0 and _length = string.Length`.
![изображение](https://user-images.githubusercontent.com/21261007/105179354-4d6aae80-5b3a-11eb-9ee0-4055acca5baa.png)

Reply:
Right, I get that.  But the code you pointed to only executes when an input variable's name is nested deeper than one level deep, or an exception occurs.  Neither of which happen during an introspection query.

Reply:
If you add a breakpoint in `ROM.ToString()` and run the `validate_core_schema` test, you will find that it only executes `ROM.ToString()` once, where the operation name ("IntrospectionQuery") is cast to a string so it can be stored in `Metrics`.

Reply:
>  But the code you pointed to only executes when an input variable's name is nested deeper than one level deep, or an exception occurs.

I wrote in the first place I saw.